### PR TITLE
httpcli

### PR DESCRIPTION
**Pull request by RoongJin-APIPlus**

### CODE REVIEW

File: httpcli/.DS_Store
Comment (Nathavee-mh) [inline httpcli/.DS_Store:1]: add this extension in git ignore

File: httpcli/httpcli
Comment (Nathavee-mh) [inline httpcli/httpcli:1]: add in git ignore

File: httpcli/server.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+			}
+
+			postCmd.Parse(os.Args[3:])
+
+			if *postJSON {
+				s := os.Args[4]
+
+				var js jsBody
+
+				err := json.Unmarshal([]byte(s), &js)
```
Comment (Nathavee-mh) [inline httpcli/server.go:169]: use json.Valid

File: httpcli/server.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		postCmd := flag.NewFlagSet("post", flag.ExitOnError)
+		postHelp := postCmd.Bool("help", false, "help")
+		postJSON := postCmd.Bool("json", false, "json")
+		deleteCmd := flag.NewFlagSet("delete", flag.ExitOnError)
+		deleteHelp := deleteCmd.Bool("help", false, "help")
+		putCmd := flag.NewFlagSet("put", flag.ExitOnError)
+		putHelp := putCmd.Bool("help", false, "help")
+
+		switch os.Args[1] {
+		case "get":
```
Comment (Gromitmugs) [inline httpcli/server.go:61]: we could get a cleaner main by calling a function in each switch case statement
the example would look like this

```go
case "get":
    util.Get(<necessary args>)
case "post":
    util.Post(<necessary args>)
...
```

File: httpcli/server.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+				var headerFlags arrayFlags
+				getCmd.Var(&queryFlags, "query", "query")
+				getCmd.Parse(os.Args[3:])
+				url := "https://" + os.Args[2] + "?"
+				for val := range queryFlags {
+					url += queryFlags[val] + "&"
+				}
+
+				fmt.Println("Request Sent to ", url[:len(url)-1])
+				client := http.Client{}
```
Comment (Gromitmugs) [inline httpcli/server.go:76]: Always add a timeout to httpclient

```
client := http.Client{
   Timeout: time.Second * 90 // 90 seconds timeout
}
```

File: httpcli/server.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+func (i *arrayFlags) Set(value string) error {
+	*i = append(*i, value)
+	return nil
+}
+
+func main() {
+	if len(os.Args) == 2 {
+		// helpCmd := flag.String("help", "", "provide http sub-commands")
```
Comment (Nathavee-mh) [inline httpcli/server.go:34]: remove unused code

File: httpcli/server.go (hunk 2 above)
Comment (Gromitmugs) [reply httpcli/server.go:61]: By doing this you will have to create a new directory and package `util` and import the functions here

**Review by Gromitmugs — CHANGES_REQUESTED** [review]
Overall the concept and fundamentals are excellent, however the code can be more organized, rather than everything in `main()` function

File: httpcli/server.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+				fmt.Println([]byte(s))
+				fmt.Println(js)
+				if err != nil {
+					fmt.Println("Wrong JSON format")
+					return
+				}
+
+				var jsonBody = []byte(s)
+
+				resp, e := http.NewRequest("POST", "https://"+os.Args[2], bytes.NewBuffer(jsonBody))
```
Comment (Nathavee-mh) [inline httpcli/server.go:125]: Add comment what os.Args[...] or assign it to variable. The variable name will tell what it is.

File: httpcli/server.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+				return
+			} else {
+				resp, e := http.NewRequest("DELETE", "https://"+os.Args[2], strings.NewReader(os.Args[3]))
+				if e != nil {
+					log.Fatal(e)
+				} else {
+					fmt.Println("DELETE command established")
+					fmt.Println(resp)
+				}
+			}
```
Comment (Nathavee-mh) [inline httpcli/server.go:154]: create a literal function or declare a new function for each http methods.

File: httpcli/server.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+)
+
+//const defaultPort = "8080"
+
+type jsBody struct {
+	Name string `json:"key" validate:"required"`
+	Age  int    `json:"age" validate:"required"`
+}
+
+type arrayFlags []string
```
Comment (Nathavee-mh) [inline httpcli/server.go:21]: this type is never use

**Review by Nathavee-mh — CHANGES_REQUESTED** [review]

**Pull request closed without merging**